src/python/pants/backend/experimental/visibility/register.py

```diff
@@ -0,0 +1,191 @@
+    ValidatedDependencies,
+    ValidateDependenciesRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+
+
+class BuildFileVisibilityImplementationRequest(BuildFileDependencyRulesImplementationRequest):
+    pass
+
+
+class VisibilityValidateFieldSet(FieldSet):
+    required_fields = (Dependencies,)
+
+
+class VisibilityValidateDependenciesRequest(ValidateDependenciesRequest):
+    field_set_type = VisibilityValidateFieldSet
+
+
+@dataclass(frozen=True)
+class VisibilityRule(DependencyRule):
+    action: DependencyRuleAction
+    pattern: str
+
+    @classmethod
+    def parse(cls, rule: str) -> VisibilityRule:
+        if rule.startswith("!"):
+            action = DependencyRuleAction.DENY
+            pattern = rule[1:]
+        elif rule.startswith("?"):
+            action = DependencyRuleAction.WARN
+            pattern = rule[1:]
+        else:
+            action = DependencyRuleAction.ALLOW
+            pattern = rule
+        return cls(action, pattern)
+
+    def match(self, path: str, relpath: str) -> bool:
+        pattern = relpath if self.pattern == "." else self.pattern
+        if pattern.startswith("./"):
+            pattern = relpath + pattern[1:]
+        return fnmatch(path, pattern)
+
+
+VisibilityRules = Tuple[VisibilityRule, ...]
+
+
+@dataclass(frozen=True)
+class BuildFileVisibilityRules(BuildFileDependencyRules):
+    default: DependencyRuleAction
+    all: DependencyRules
+    targets: Mapping[str, DependencyRules]
+
+    @classmethod
+    def create(
+        cls,
+        default: DependencyRuleAction = DependencyRuleAction.ALLOW,
+        all: Iterable[str | DependencyRule] = (),
+        targets: Mapping[str, Iterable[str | DependencyRule]] = {},
+    ) -> BuildFileVisibilityRules:
+        return cls(
+            default,
+            cls.parse_visibility_rules(all),
+            FrozenDict(
+                {
+                    type_alias: cls.parse_visibility_rules(rules)
+                    for type_alias, rules in targets.items()
+                }
+            ),
+        )
+
+    @classmethod
+    def parse_visibility_rules(cls, rules: Iterable[str | DependencyRule]) -> DependencyRules:
+        return tuple(
+            VisibilityRule.parse(rule) if isinstance(rule, str) else rule for rule in rules
+        )
+
+    def get_rules(self, type_alias: str) -> VisibilityRules:
+        if type_alias in self.targets:
+            return cast(VisibilityRules, self.targets[type_alias])
+        else:
+            return cast(VisibilityRules, self.all)
+
+    def get_action(self, type_alias: str, path: str, relpath: str) -> DependencyRuleAction:
+        """Get applicable rule for target type from `path`.
+
+        The rules are declared in `relpath`.
+        """
+        for visibility_rule in self.get_rules(type_alias):
+            if visibility_rule.match(path, relpath):
+                return visibility_rule.action
+        return self.default
```

Comment:
I don't see where "all" is applied?

Reply:
`self.get_rules(type_alias)` returns `all` rules for the types that doesn't have a target specific list of rules. As mentioned in my reply to the thread in that method I think it may make sense to always return `all` rules in addition to any target specific rules.

Reply:
Ah, this is what confused me - I expected `all` to always apply, even if there is also a rule for the type_alias.  The behavior you had before was more like `default`...

Reply:
Stale thread..